### ISSUE TITLE
git lfs does not work on windows bash.exe when used with unicode paths

### ISSUE THREAD

**Issue by J0Nes90** [git-lfs/git-lfs#4227]
**Describe the bug**
This is a follow up from https://stackoverflow.com/questions/63691374/git-lfs-checkout-fails-on-windows-with-unicode-path-error-createfile
I am running Windows 10 Pro in a virtual machine with nothing but Git for Windows installed. I can reproduce the issue on a physical machine as well, however.
Running any LFS command that modifies the working directory fails on bash.exe when the path to the repository contains Unicode characters. This can also be reproduced when executing a git lfs pull through GitKraken (which also seems to invoke bash.exe).

**To Reproduce**
Run the following shell script on bash.exe and compare it to git-bash.exe
```
createRepo()
{
  directory=$1
  mkdir $directory
  cd $directory
  git init
  git lfs install --local
  git config --local user.email "test@mail.com"
  git config --local user.name "test"
  git lfs track bin.dat
  git add .gitattributes
  git commit -m "attribute file"

  file="bin.dat"
  echo "qwertyuiop" > "bin.dat"

  git add .
  git commit -m "added binary file"

  cd ..
}

mkdir lfsissue
cd lfsissue

echo "create repo in path with ascii characters only"
createRepo test

echo "create repo in path with unicode characters"
createRepo äöü
```

Compare the output from bash.exe
```
$ sh issue.sh
create repo in path with ascii characters only
Initialized empty Git repository in C:/Users/VM/Desktop/lfsissue/test/.git/
Updated git hooks.
Git LFS initialized.
Tracking "bin.dat"
[master (root-commit) 88018d1] attribute file
 1 file changed, 1 insertion(+)
 create mode 100644 .gitattributes
[master 9d69353] added binary file
 1 file changed, 3 insertions(+)
 create mode 100644 bin.dat
create repo in path with unicode characters
Initialized empty Git repository in C:/Users/VM/Desktop/lfsissue/äöü/.git/
Error: CreateFile C:\Users\VM\Desktop\lfsissue\���: The system cannot find the file specified.
Updated git hooks.
Git LFS initialized.
Error: CreateFile C:\Users\VM\Desktop\lfsissue\���: The system cannot find the file specified.
This operation must be run in a work tree.
fatal: pathspec '.gitattributes' did not match any files
On branch master

Initial commit

nothing to commit (create/copy files and use "git add" to track)
warning: LF will be replaced by CRLF in bin.dat.
The file will have its original line endings in your working directory
Error: CreateFile C:\Users\VM\Desktop\lfsissue\���: The system cannot find the file specified.
[master (root-commit) d316732] added binary file
 1 file changed, 1 insertion(+)
 create mode 100644 bin.dat
```
to the correct and expected output of git-bash.exe
```
$ sh issue.sh
create repo in path with ascii characters only
Initialized empty Git repository in C:/Users/VM/Desktop/lfsissue/test/.git/
Updated git hooks.
Git LFS initialized.
Tracking "bin.dat"
[master (root-commit) 5816b6c] attribute file
 1 file changed, 1 insertion(+)
 create mode 100644 .gitattributes
[master b5276fa] added binary file
 1 file changed, 3 insertions(+)
 create mode 100644 bin.dat
create repo in path with unicode characters
Initialized empty Git repository in C:/Users/VM/Desktop/lfsissue/├ñ├Â├╝/.git/
Updated git hooks.
Git LFS initialized.
Tracking "bin.dat"
[master (root-commit) d4c2fba] attribute file
 1 file changed, 1 insertion(+)
 create mode 100644 .gitattributes
[master 7bf822e] added binary file
 1 file changed, 3 insertions(+)
 create mode 100644 bin.dat

```

**Output of `git lfs env`**
git-lfs/2.11.0 (GitHub; windows amd64; go 1.14.2; git 48b28d97)
git version 2.28.0.windows.1

LocalWorkingDir=
LocalGitDir=
LocalGitStorageDir=
LocalMediaDir=lfs\objects
LocalReferenceDirs=
TempDir=lfs\tmp
ConcurrentTransfers=8
TusTransfers=false
BasicTransfersOnly=false
SkipDownloadErrors=false
FetchRecentAlways=false
FetchRecentRefsDays=7
FetchRecentCommitsDays=0
FetchRecentRefsIncludeRemotes=true
PruneOffsetDays=3
PruneVerifyRemoteAlways=false
PruneRemoteName=origin
LfsStorageDir=lfs
AccessDownload=none
AccessUpload=none
DownloadTransfers=basic,lfs-standalone-file
UploadTransfers=basic,lfs-standalone-file
GIT_EXEC_PATH=C:/Program Files/Git/mingw64/libexec/git-core
git config filter.lfs.process = "git-lfs filter-process"
git config filter.lfs.smudge = "git-lfs smudge -- %f"
git config filter.lfs.clean = "git-lfs clean -- %f"

**Comment by bk2204** [comment]
Hey,

When you say `bash.exe`, are you are talking about the Windows Subsystem for Linux?  And if so, are you using the Git for Windows version of Git and the Windows version of Git LFS in WSL?  What distro are you using?  Can you provide the output of the following commands in both environments?

```
$ echo "$PATH"
$ which git
$ which git-lfs
$ grep createRepo script.sh | od -tx1
```

I'm not seeing this in either case in Windows, either under WSL or Git Bash, or on Linux.  I suspect what's happening is that your script is encoded in a character set other than UTF-8 and that's causing problems, but I'm not 100% sure.

**Comment by J0Nes90** [comment]
Hey,

no, I am referring to the bash.exe that is provided through the "[Git for Windows](https://gitforwindows.org/)" installation. 
`C:\Program Files\Git\bin\bash.exe`

The machine is vanilla windows 10, nothing installed but Git for Windows. Hence, no distro / Linux / WSL involved at all.
Running those commands on bash.exe and git-bash.exe yield the very same results, just the PATH is different:

which git
`/mingw64/bin/git`

which git-lfs
`/mingw64/bin/git-lfs`

grep createRepo script.sh | od -tx1
```
0000000 63 72 65 61 74 65 52 65 70 6f 28 29 0d 0a 63 72
0000020 65 61 74 65 52 65 70 6f 20 74 65 73 74 0d 0a 63
0000040 72 65 61 74 65 52 65 70 6f 20 c3 a4 c3 b6 c3 bc
0000060 0d 0a
0000062
```

_Note: The script is not the issue, I can reproduce the issue when executing those commands manually as well_

echo "$PATH" (bash.exe)
`/mingw64/bin:/usr/bin:/c/Users/VM/bin:/c/Windows/system32:/c/Windows:/c/Windows/System32/Wbem:/c/Windows/System32/WindowsPowerShell/v1.0:/c/Windows/System32/OpenSSH:/c/Users/VM/AppData/Local/Microsoft/WindowsApps:/bin:/c/Users/VM/AppData/Local/Microsoft/WindowsApps`

echo "$PATH" (git-bash.exe)
`/c/Users/VM/bin:/mingw64/bin:/usr/local/bin:/usr/bin:/bin:/mingw64/bin:/usr/bin:/c/Users/VM/bin:/c/Windows/system32:/c/Windows:/c/Windows/System32/Wbem:/c/Windows/System32/WindowsPowerShell/v1.0:/c/Windows/System32/OpenSSH:/c/Users/VM/AppData/Local/Microsoft/WindowsApps:/bin:/c/Users/VM/AppData/Local/Microsoft/WindowsApps:/usr/bin/vendor_perl:/usr/bin/core_perl`

**Comment by bk2204** [comment]
Okay, I've been able to reproduce this, and here's what I've found.  On Windows, paths are stored in UTF-16.  All of the standard Unix tools, Git and Git LFS included, expect to use and produce UTF-8, since they use 8-bit strings.  As part of the process of looking up the repository root on Windows, we invoke `cygpath -w` to get the canonical Windows path.  That's because Git LFS is a Go binary, and Go, unlike most other languages, doesn't use the system's libc, so we need to turn any Cygwin or MINGW (e.g., slash-formatted) paths into a native Windows path we can use.

This normally works great, except that `cygpath` produces output in some non-UTF-8 format.  This is probably because when you use `git-bash.exe`, it sets up the locale properly, but when you use `bash.exe`, it does not set up the locale properly, so `cygpath` doesn't know that you want UTF-8.  It looks, from trying to use a Cyrillic character, that the character set it is using is ISO-8859-1.

This is bizarre, because there's literally no way to represent every path on Windows using only ISO-8859-1, and as such, that's never a sensible default on Windows.  (Unix, on the other hand, has arbitrary bytes as paths, so encoding is not a problem there.)  I would argue this is a bug in `cygpath`.  However, because it's unlikely that we'll actually get `cygpath` fixed in any reasonable amount of time, I'm okay with putting a change in here to fix it.  I have such a change and I'll add a test.

In the mean time, if you set `LC_ALL=C.UTF-8` (or another UTF-8 locale) in your environment, and things should work.
